### ISSUE TITLE
Use different datastore locations for local runs

### ISSUE THREAD

**Issue by chanseokoh** [GoogleCloudPlatform/google-cloud-eclipse#937]
Even after fixing #393, we cannot run two local servers simultaneously, since subsequent servers will try to use the same datastore that the first server is using.

`OperationalError: database is locked`

I believe providing different paths to dev_appserver.py via `--storage_path` or `--datastore_path` would fix this.

**UPDATE: we won't fix the above. This issue is to fix [the comment](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/937#issuecomment-257996776) below.**

**Comment by chanseokoh** [comment]
Another problem we see when we hit this bug is that, the processes of the subsequent servers will not terminate, while in Eclipse, they are marked as "stopped." Terminating Eclipse doesn't terminate them either. Uses have to kill the processes manually.

**Comment by elharo** [comment]
Is this something we're actually doing within Eclipse? Or is the problem only if we run one server in Eclipse and one at the command line?

**Comment by chanseokoh** [comment]
We can hit this bug in Eclipse by creating and running two servers. (Reproducible only after fixing #393, which allows us to run two servers with different ports.)

**Comment by patflynn** [comment]
Why are you running two dev appserver.py's concurrently?

**Comment by chanseokoh** [comment]
@patflynn The Eclipse WTP framework allows to create and run any number of servers of any type (not just our dev app server). Users can do this and will hit this bug.

We can let the second instance of server to fail as in the current state, if that's our policy. However, we still need to do fix https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/937#issuecomment-257996776 even in this case. The server instance goes into a broken state, and users have to manually kill it.

**Comment by patflynn** [comment]
IJ has a dialog to 'disconnect i.e. kill' a running service if you try to
quit while it's on.  Does Eclipse have something similar?

I don't think we should fix the lock issue in the IDE.  I don't think the
dev_appserver.py is designed to run multiple instances on the same machine.

**Comment by chanseokoh** [comment]
> IJ has a dialog to 'disconnect i.e. kill' a running service

Eclipse can does a similar thing. What I am saying is that our server goes into a broken state when we hit this, and thus trying to disconnect and kill does not work. We still need to fix that bug if we won't fix the lock issue. I'll change the title of this issue to reflect that. We won't fix the lock issue.

**Comment by briandealwis** [comment]
> @chanseokoh: I believe providing different paths to dev_appserver.py via `—storage_path` or `—datastore_path` would fix this.

Makes sense to me.

Where do these default to normally?

We could default to setting these values to the server-managed temp directory (the same place where we publish the modules to).  That temp directory has the same lifetime as the server instance itself.

Perhaps we should default all of these paths:
  - storage_path
  - blobstore_path
  - datastore_path
  - prospective_search_path
  - search_indexes_path
  - logs_path

**Comment by chanseokoh** [comment]
@briandealwis I guess our decision is now to let the second server instance to fail with the database lock error. (Please see the discussions above.)

In any case, dev_appserver.py creates a temp directory under `/tmp` on my machine. You could run dev_appserver.py and observe this. I think `storage_path` can cover of at least some of other paths. (For example, unless you set `datastore_path` explicitly, it uses `storage_path`.)

**Comment by briandealwis** [comment]
> @patflynn: I don't think the dev_appserver.py is designed to run multiple instances on the same machine.

Between storage locations and default ports, it will be difficult.

**Comment by elharo** [comment]
This sounds like a nice to fix if it's easy, but right now maybe we have higher priorities. If users scream that things break when they multiple servers we can revisit.

**Comment by patflynn** [comment]
I don't think you'd  necessarily want to put these files in a short-lived
tmp directory. for instance the datastore emulator has the option of
keeping it's persisted entities between server runs.

**Comment by briandealwis** [comment]
I might have been better calling that directory 'a server-instance-specific state directory'.  It's not short-lived, but rather bounded by the lifetime of the server instance.  The user can start and stop the server multiple times and the directory continues to exist using the user deletes that server instance.  For example, here are three server instances I have defined in my test workspace:

![screen shot 2016-11-03 at 2 57 04 pm](https://cloud.githubusercontent.com/assets/202851/19980796/d4a13ec8-a1d5-11e6-88a1-e549c7a6d199.PNG)

The directories are stored in a non-visible location within the Eclipse workspace's state directory.

**Comment by chanseokoh** [comment]
Per @briandealwis's suggestion, I reverted the title back to the original and instead created #942 for the unterminatable dev_appserver.

I believe our decision is not to fix this database lock issue, so I'll close this. If you have any concern, feel free to raise it.

**Comment by briandealwis** [comment]
The _devappserver_ creates its the various DBs under the _storage_path_ directory. This directory can be specified with the `—storage_path` argument.  If unspecified, then _devappserver_ seems to use `$TMP/appengine.APPID.LOGIN` (on OS X and probably Linux), where _APPID_ is from the _appengine-web.xml_'s `<application>` tag or the `-A appId` flag, and _LOGIN_ is the active user.  In the absence of an _APPID_, the _devappserver_ uses _None_.  So I see the _devappserver_ creating the storage path in `$TMP/appengine.None.bsd`.

Since our WTP server instances have unique names, we could provide them as the `-A APPID`, in the absence of an _APPID_.

**Comment by briandealwis** [comment]
Or perhaps we should launch with `—storage_path` pointing within that _server-instance-specific state directory_.
